src/network/networkd-route.c

```diff
@@ -87,20 +87,94 @@ static const char * const route_table_table[] = {
         [RT_TABLE_LOCAL]   = "local",
 };
 
-DEFINE_STRING_TABLE_LOOKUP(route_table, int);
+DEFINE_PRIVATE_STRING_TABLE_LOOKUP(route_table, int);
 
-#define ROUTE_TABLE_STR_MAX CONST_MAX(DECIMAL_STR_MAX(int), STRLEN("default") + 1)
-static const char *format_route_table(int table, char *buf, size_t size) {
+int route_table_from_string_full(const Manager *m, const char *s, uint32_t *ret) {
+        uint32_t t;
+        int r;
+
+        assert(s);
+        assert(ret);
+
+        r = route_table_from_string(s);
+        if (r >= 0) {
+                *ret = (uint32_t) r;
+                return 0;
+        }
+
+        if (m) {
+                t = PTR_TO_UINT32(hashmap_get(m->route_table_numbers_by_name, s));
+                if (t != 0) {
+                        *ret = t;
+                        return 0;
+                }
+        }
+
+        r = safe_atou32(s, &t);
+        if (r < 0)
+                return r;
+
+        if (t == 0)
+                return -ERANGE;
+
+        *ret = t;
+        return 0;
+}
+
+int route_table_to_string_full(const Manager *m, uint32_t table, char **ret) {
+        _cleanup_free_ char *str = NULL;
         const char *s;
-        char *p = buf;
+
+        assert(ret);
+
+        if (table == 0)
+                return -EINVAL;
 
         s = route_table_to_string(table);
-        if (s)
-                strpcpy(&p, size, s);
-        else
-                strpcpyf(&p, size, "%d", table);
+        if (s) {
+                str = strdup(s);
+                if (!str)
+                        return -ENOMEM;
 
-        return buf;
+                *ret = TAKE_PTR(str);
+                return 0;
+        }
+
+        if (m) {
+                s = hashmap_get(m->route_table_names_by_number, UINT32_TO_PTR(table));
+                if (s) {
+                        /* Currently, this is only used in debugging logs. To not confuse any bug
+                         * reports, let's include the table number. */
+                        if (asprintf(&str, "%s(%" PRIu32 ")", s, table) < 0)
+                                return -ENOMEM;
+
+                        *ret = TAKE_PTR(str);
+                        return 0;
+                }
+        }
+
+        if (asprintf(&str, "%" PRIu32, table) < 0)
+                return -ENOMEM;
+
+        *ret = TAKE_PTR(str);
+        return 0;
+}
+
+void route_tables_free(Manager *m) {
+        char *name;
+
+        assert(m);
+
+        /* route table names may be shared by the two hashmaps. So, let's first clear all entries in
+         * route_table_names_by_number, and then free the two hashmaps. */
+
+        while ((name = hashmap_steal_first(m->route_table_names_by_number))) {
+                hashmap_remove(m->route_table_numbers_by_name, name);
+                free(name);
+        }
+
+        m->route_table_names_by_number = hashmap_free(m->route_table_names_by_number);
+        m->route_table_numbers_by_name = hashmap_free(m->route_table_numbers_by_name);
 }
```

Comment:
```
m->route_table_names_by_number = hashmap_free(m->route_table_names_by_number);
m->route_table_numbers_by_name = hashmap_free_free_key(m->route_table_numbers_by_name)
```

Then the above logic can be dropped.

Reply:
That's just for safety. An entry may be added in only one hashmap. Please see `config_parse_table_names()`. We have
```
if (r > 0 || k > 0)
    TAKE_PTR(name);
```
So, at least one of `hashmap_ensure_put()` may return 0.

Reply:
OK. By reconsidering your comments, I dropped the logic. Thanks.